src/org/etools/j1939_84/controllers/part09/Part09Step22Controller.java

```diff
@@ -54,12 +57,40 @@
     @Override
     protected void run() throws Throwable {
         // 6.9.22.1.a. Global DM2 [(send Request (PGN 59904) for PGN 65227 (SPNs 1213-1215, 1706, and 3038)]).
-        // 6.9.22.2.a. (if supported) Fail if any ECU does not report MIL off or MIL not supported. See Section A.8 for
-        // allowed values.
-        // 6.9.22.2.b. (if supported) Fail if any OBD ECU reports a previously active DTC.
+        var globalPackets = getDiagnosticMessageModule().requestDM2(getListener())
+                                                        .getPackets()
+                                                        .stream()
+                                                        .peek(this::save)
+                                                        .filter(p -> getDataRepository().isObdModule(p.getSourceAddress()))
+                                                        .peek(p -> {
+                                                            // 6.9.22.2.a. (if supported) Fail if any ECU does not
+                                                            // report
+                                                            // MIL off or MIL not supported.
+                                                            // See Section A.8 for allowed values.
+                                                            if (isNotOff(p.getMalfunctionIndicatorLampStatus())) {
```

Comment:
This logic isn't correct.

I think you need to do this outside of this stream.

Reply:
Actually, logic was right verbiage was wrong.  Fixed.

Reply:
It's still wrong.

You need to cycle through _all_ of the packets and determine if any ECU has MIL off (or alt off).
If no module has MIL set to off or alt off, you need to write a failure.

Reply:
I read that wrong.  But the logic is still not correct.

Fail if the a module reports `(mil != OFF && mil != NOT_SUPPORTED)` should be the logic

Reply:
```
.filter(p -> {
var mil = p.getMalfunctionIndicatorLampStatus();
return isNotOff(mil) && mil != NOT_SUPPORTED
})
```